File: Langauge/set/compiler/java/src/main/java/set/antlr4/SymbolVisitor.java
package set.antlr4;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.TerminalNode;

import set.antlr4.SETParser.FunctionDeclContext;

public class SymbolVisitor extends SETBaseVisitor<SETValue> {
    Map<String, Function> functions;
    
    public SymbolVisitor(Map<String, Function> functions) {
        this.functions = functions;
    }
    
    @Override
    public SETValue visitFunctionDecl(FunctionDeclContext ctx) {
        List<TerminalNode> params = ctx.idList() != null ? ctx.idList().Identifier() : new ArrayList<TerminalNode>(); 
        ParseTree block = ctx.block();
        String id = ctx.Identifier().getText() + params.size();
        functions.put(id, new Function(id, params, block));
        return SETValue.VOID;
    }
}

File: Music/Langauge/set/compiler/cpp/main/cpp/set/antlr4/SymbolVisitor.cpp
#include "SymbolVisitor.h"
#include "TLValue.h"
#include "Function.h"

namespace tl
{
	namespace antlr4
	{
		using org::antlr::v4::runtime::tree::ParseTree;
		using org::antlr::v4::runtime::tree::TerminalNode;
		using tl::antlr4::TLParser::FunctionDeclContext;

		SymbolVisitor::SymbolVisitor(std::unordered_map<std::wstring, Function*> &functions)
		{
			this->functions = functions;
		}

		TLValue *SymbolVisitor::visitFunctionDecl(FunctionDeclContext *ctx)
		{
			std::vector<TerminalNode*> params = ctx->idList() != nullptr ? ctx->idList().Identifier() : std::vector<TerminalNode*>();
			ParseTree *block = ctx->block();
			std::wstring id = ctx->Identifier().getText() + params.size();
			functions[id] = new Function(id, params, block);
			return TLValue::VOID;
		}
	}
}
